xxxx-generic-js-field.md

```diff
@@ -0,0 +1,170 @@
+
+The plan is best described using this mock OCaml code:
+
+```ocaml
+module Js = Js_of_ocaml.Js
+
+(* interfaces that snarky expects and returns *)
+module type Field_intf = sig
+  type t
+
+  val add : t -> t -> t
+
+  val sub : t -> t -> t
+
+  val mul : t -> t -> t
+
+  val zero : t
+
+  val one : t
+end
+
+module type Backend_intf = sig
+  module Field : Field_intf
+end
+
+module type Snarky_intf = sig
+  module Field : Field_intf
+
+  module Constraint : sig
+    type t
+
+    val r1cs : Field.t -> Field.t -> Field.t -> t
+
+    val equal : Field.t -> Field.t -> t
+  end
+
+  val assert_ : Constraint.t -> unit
+end
+
+module Snarky (Backend : Backend_intf) = struct
+  module Field = Backend.Field
+
+  module Constraint = struct
+    type t = unit
+
+    let r1cs _x _y _z : t = ()
+
+    let equal _x _y : t = ()
+  end
+
+  let assert_ (_c : Constraint.t) : unit = ()
+end
+
+(* generic native field module, reimplemented in JS *)
+type bigint = Js.Unsafe.any Js.t
+
+type field = bigint ref
+
+type js_field_module =
+  < add : (field -> field -> field) Js.readonly_prop
+  ; sub : (field -> field -> field) Js.readonly_prop
+  ; mul : (field -> field -> field) Js.readonly_prop
+  ; zero : field Js.readonly_prop
+  ; one : field Js.readonly_prop >
+  Js.t
+
+external create_js_field : field -> js_field_module = "create_field"
+
+(* make existing snarky-bindings a functor from snarky *)
+module Snarky_bindings (Impl : Snarky_intf) = struct
+  module Field' = struct
+    let assert_equal x y = Impl.assert_ (Impl.Constraint.equal x y)
+
+    let assert_mul x y z = Impl.assert_ (Impl.Constraint.r1cs x y z)
+  end
+end
+
+(* instantiating snarky and snarky-bindings with JS field module, and return something usable in JS *)
+let create_field (modulus : field) =
+  (* call generic JS field constructor *)
+  let js_field = create_js_field modulus in
+  (* instantiate backend *)
+  let module Backend = struct
+    module Field = struct
+      type t = field
+
+      let add = js_field##.add
+
+      let sub = js_field##.sub
+
+      let mul = js_field##.mul
+
+      let zero = js_field##.zero
+
+      let one = js_field##.one
+    end
+  end in
+  (* instantiate snarky *)
+  let module Snarky' = Snarky (Backend) in
+  (* instantiate snarky-bindings *)
+  let module Snarky_bindings' = Snarky_bindings (Snarky') in
+  (* return for JS use *)
+  object%js
+    method assertEqual = Snarky_bindings'.Field'.assert_equal
+
+    method assertMul = Snarky_bindings'.Field'.assert_mul
+  end
+```
+
+### Rust side: instantiating core Kimchi methods
+
+TODO - does Rust enable us to dynamically instantiate traits (or whatever) as function return values?
```

Comment:
https://doc.rust-lang.org/rust-by-example/trait/dyn.html , but it's "slow" since there is a few pointer jumps vs the static https://doc.rust-lang.org/rust-by-example/trait/impl_trait.html . I don't think it will be even remotely close to the bottleneck here though.

Reply:
But that example picks between 2 traits that both have been declared already, instead of _creating_ a trait depending on a parameter like we can do in OCaml and JS

Reply:
Ah okay I get it, typically I use a “record of closures” to get that to work in these kinds of languages though I haven’t tried it in Rust.

The idea is for all the functions of the trait you add closures to the struct and the implementations of the trait is just calling the closures. Then you can, at runtime, swap implementations. I don’t see a reason why it won’t work in rust if you wrap things in enough boxes, but I’m not a Rust expert.